API updates and new endpoint

mpretty-cyro opened the pull request:
• Added a Dockerfile to simplify local testing
• Added a 'release_assets' table to store links to release assets
• Added an 'account_version_checks' for version check auth requests
• Added 'session_token_stats' to store daily records of token stats
• Added a 'token_info' endpoint to return token info to the caller
• Updated the 'release_versions' table to include more release information
• Updated the 'session_version' API to take optional (blinded) auth information and return more release information

mpretty-cyro committed:
WIP for API updates and new endpoint

• Added a Dockerfile to simplify local testing
• Added a 'release_assets' table to store links to release assets
• Added an 'account_version_checks' for version check auth requests
• Added 'session_token_stats' to store daily records of token stats
• Added a 'token_info' endpoint to return token info to the caller
• Updated the 'release_versions' table to include more metadata
• Updated the 'session_version' API to take optional auth information and return more release metadata

mpretty-cyro committed:
Updated data structures and got fetching logic working

mpretty-cyro committed:
Removed test token data

mpretty-cyro committed:
Updated API docs, tweaked the token_info endpoint

jagerman commented on fileserver/crypto.py:
```diff
@@ -1,6 +1,7 @@
 import os
-import pyonionreq.junk
+import pyonionreq
+from pyonionreq import junk
```
We should get rid of this dependency here: `onionreq` was a half-baked helper lib that has been surpassed by functionality in `python3-session-util` (which are (some) python bindings for libsession-util).

The replacement is quite simple: this is the commit that updated SOGS: https://github.com/oxen-io/session-pysogs/pull/216/commits/7eecddd14611cfb82f03dcfc9e1628ce14d92727 (although the `import` there is missing a line: it should be `from session_util.onionreq import OnionReqParser`)

mpretty-cyro replied:
I tried to swap this out (have all the changes locally) but it ran into the below issue - did I need to add `libsession-util` as a submodule or dependency as well somehow? (I couldn't see this being done in the PySOGS branch so wasn't sure how it should be done 😕)
```
The following packages have unmet dependencies:
  python3-session-util : Depends: libsession-util1.1.0 but it is not installable
```

mpretty-cyro replied:
This has been updated now

jagerman commented on schema.pgsql:
```diff
@@ -16,14 +16,61 @@ CREATE INDEX files_expiry ON files(expiry);
+    blinded_id varchar(66) NOT NULL,
+    platform varchar(25) NOT NULL,
+    timestamp TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+    CONSTRAINT unique_blinded_id_platform_timestamp UNIQUE (blinded_id, platform, timestamp)
```
This unique index seems unlikely to do anything since `timestamp = NOW()` is pretty much always going to be a different value.

mpretty-cyro replied:
Happy to remove this - I wasn't sure how precise the `TIMESTAMP` time was so wasn't sure if we could get collisions if two devices happen to hit the endpoint at the same time

jagerman commented on schema.pgsql:
```diff
@@ -16,14 +16,61 @@ CREATE INDEX files_expiry ON files(expiry);
+-- Abritrary values at the (approx) time this was written; this don't really matter as
+-- they'll get updated within a first few seconds of initial startup.
+INSERT INTO session_token_stats (maximum_supply, sent_per_node, staking_reward_pool) VALUES (240000000, 15000, 40000000);
+INSERT INTO session_token_history (current_value, circulating_supply, total_nodes, updated) VALUES (0.099002, 30400000, 2194, 68297852, '2024-06-14Z');
```
I couldn't find anything here that actually updates these (as the comment suggests happens somewhere)

mpretty-cyro replied:
So these tables are used for the new `/token_info` endpoint (which we have in-progress designs for - I think I need to make minor changes to the API/structure based on the latest designs anyway)

For this one I was asked to stub-out an endpoint to populate the screen (and do as much work in advance as I could) but, since the token doesn't exist yet, creating a script to populate these tables wasn't really something I could do - I was envisioning something like what we have in `cleanup.py` which just fetches token stats once per day from somewhere and stores them in here

mpretty-cyro replied:
These values are actually also test data that I should have removed - will remove them from the next commit

I'm just wondering whether it'd be better for the `/token_info` to just return 0-ed values if there is nothing in the database (at the moment it returns a `502` error) 🤔

jagerman commented on schema.pgsql:
```diff
@@ -16,14 +16,61 @@ CREATE INDEX files_expiry ON files(expiry);
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-desktop', 'v1.7.3', NULL, '2021-10-14Z');
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-android', '1.11.11', NULL, '2021-10-14Z');
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-ios', '1.11.17', NULL, '2021-10-14Z');
+
```
There's some (undeclared) multi-column foreign keys here which we could normalize to clean up the structure a bit.  I played around with it a bit and came up with this:

```SQL
CREATE TABLE projects (
    id BIGSERIAL PRIMARY KEY,
    name varchar(50) NOT NULL,
    updated timestamp with time zone NOT NULL DEFAULT NOW()
);

CREATE TABLE releases (
    id BIGSERIAL PRIMARY KEY,
    project BIGINT NOT NULL REFERENCES projects(id),
    prerelease BOOLEAN NOT NULL DEFAULT FALSE,
    version_code BIGINT NOT NULL,
    url varchar(255) NOT NULL,
    name varchar(255),
    notes text,
    UNIQUE(project, version_code)
);

CREATE TABLE release_assets (
    release BIGINT NOT NULL REFERENCES releases(id),
    name varchar(255) NOT NULL,
    url varchar(255) NOT NULL
);
CREATE INDEX ON release_assets(release);

CREATE VIEW versions AS
    SELECT
        releases.id as id,
        projects.name as proj_name,
        version_code,
        version_code / 1000000 || '.' || version_code % 1000000 / 1000 || '.' || version_code % 1000 AS version,
        prerelease,
        url,
        releases.name AS name,
        notes
    FROM releases JOIN projects ON releases.project = projects.id;

CREATE VIEW release_versions AS SELECT * FROM versions WHERE NOT prerelease;
CREATE VIEW prerelease_versions AS SELECT * FROM versions WHERE prerelease;
```

The `version_code` would be a numeric representation of the version, e.g. we'd insert 2.3.4 as `2003004` (and the VIEW, above, converts it back to the dotted form upon querying).

The whole thing then gets populated like this:

```SQL
insert into projects (name) values ('abc');
insert into releases (project, version_code, url, name, notes) values ((select id from projects where name = 'abc'), 2003004, 'http://example.com/download/me', 'Exciting new 2.3.4 update with 26% more hokeypokey', '- Added 20% more hokeypokey
- Added 10% even more hokeypokey
- Removed 5.5% hokeypokey');
insert into releases (project, prerelease, version_code, url, name, notes) values ((select id from projects where name = 'abc'), TRUE, 2003005, 'http://example.com/download/me/again', 'Less exciting 2.3.5 update with only 3% more hokeypokey', '- Experimental new 3% hokeypokey boost');
insert into releases (project, version_code, url, name, notes) values ((select id from projects where name = 'abc'), 2003006, 'http://example.com/download/me/again', 'Less exciting 2.3.6 update with only 3% more hokeypokey', '- Final release of new 3% hokeypokey boost');
```

And querying: looks like this:
```sql
=> select * from release_versions ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                   notes                    |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
|  3 | abc       |      2003006 | 2.3.6   | f          | http://example.com/download/me/again | Less exciting 2.3.6 update with only 3% more hokeypokey | - Final release of new 3% hokeypokey boost |
|  1 | abc       |      2003004 | 2.3.4   | f          | http://example.com/download/me       | Exciting new 2.3.4 update with 26% more hokeypokey      | - Added 20% more hokeypokey               +|
|    |           |              |         |            |                                      |                                                         | - Added 10% even more hokeypokey          +|
|    |           |              |         |            |                                      |                                                         | - Removed 5.5% hokeypokey                  |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
(2 rows)

=> select * from prerelease_versions ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                 notes                  |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
|  2 | abc       |      2003005 | 2.3.5   | t          | http://example.com/download/me/again | Less exciting 2.3.5 update with only 3% more hokeypokey | - Experimental new 3% hokeypokey boost |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
(1 row)

=> select * from release_assets where release = 3;
+---------+---------+-------------------------------------------+
| release |  name   |                    url                    |
+---------+---------+-------------------------------------------+
|       3 | linux   | http://example.com/abc-2.3.6-linux.tar.gz |
|       3 | windows | http://example.com/abc-2.3.6-win.zip      |
+---------+---------+-------------------------------------------+
(2 rows)
```

Aside from better normalization in the database, this also simplifies things by no longer needing to worry about delete old release data.  Not deleting previous release data actually enables a nice capability in that if someone queries the latest version and passes their own version, we can include cumulative release notes for all the versions since the version they are currently on, by querying:
```sql
=> select * from release_versions WHERE version_code > 2003004 ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                   notes                    |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
|  4 | abc       |      3000000 | 3.0.0   | f          | http://example.com/3.0.0             | 3.0.0 release                                           | - New release to make the number bigger    |
|  3 | abc       |      2003006 | 2.3.6   | f          | http://example.com/download/me/again | Less exciting 2.3.6 update with only 3% more hokeypokey | - Final release of new 3% hokeypokey boost |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
(2 rows)
```

mpretty-cyro replied:
Yea nice, I'll make these changes now - since I don't have a lot of Python experience and hadn't touched the File Server code before I was trying to minimise the changes I made 😁

jagerman commented on schema.pgsql:
```diff
@@ -16,14 +16,61 @@ CREATE INDEX files_expiry ON files(expiry);
+
+CREATE INDEX release_assets_project_version ON release_assets(project, version);
+
+INSERT INTO release_assets (project, version, name, url) VALUES ('oxen-io/session-desktop', 'v1.7.3', 'Test name', 'github.com');
```
I don't think this initial row is needed, as there can simply be no assets associated with a given release.

mpretty-cyro replied:
I think this was test data I forget to remove 😓 

Have removed now

jagerman commented on fileserver/routes.py:
```diff
@@ -42,6 +56,101 @@ def generate_file_id(data):
+            http.BAD_REQUEST, "Invalid authentication: X-FS-Pubkey is not a valid 66-hex digit id"
+        )
+
+    if pk[0] not in (0x07,):
```
```suggestion
    if pk[0] != 0x07:
```

mpretty-cyro replied:
Updated

jagerman commented on fileserver/routes.py:
```diff
@@ -42,6 +56,101 @@ def generate_file_id(data):
+        sig_in = utils.decode_hex_or_b64(sig_in, 64)
+    except Exception:
+        abort_with_reason(
+            http.BAD_REQUEST, "Invalid authentication: X-FS-Signature is not base64[88]"
```
```suggestion
            http.BAD_REQUEST, "Invalid authentication: X-FS-Signature is not base64[86]"
```
(Only 86 b64 chars in a 64-byte value; there *can* be two extra padding bytes to pad it out to 88, but they aren't required).

mpretty-cyro replied:
Good catch - updated

jagerman commented on fileserver/routes.py:
```diff
@@ -42,6 +56,101 @@ def generate_file_id(data):
+
+    # If we were missing one of the auth headers and don't require auth then just return a 'None'
+    if missing > 0 and not required:
+        return None
```
The case where `0 < missing < 3` case is a bit weird: the client is passing *some* of the auth headers but not all.  I think we should probably raise an error in such a case:
```suggestion
    if missing:
        if required or missing < 3:
            abort_with_reason(
                http.BAD_REQUEST, "Invalid authentication: one or more X-FS-* auth headers is missing")
        else:
            return None
```

mpretty-cyro replied:
Yep fair enough - updated

jagerman commented on Dockerfile:
```diff
@@ -0,0 +1,156 @@
+RUN curl -so /etc/apt/trusted.gpg.d/oxen.gpg https://deb.oxen.io/pub.gpg && \
+    echo "deb https://deb.oxen.io $(lsb_release -sc) main" | tee /etc/apt/sources.list.d/oxen.list && \
+    apt-get update && apt-get install -y \
+    python3-pyonionreq \
```
See other comment about swapping this to python3-session-util

mpretty-cyro replied:
This has been updated now

jagerman commented on Dockerfile:
```diff
@@ -0,0 +1,156 @@
+    echo "deb https://deb.oxen.io $(lsb_release -sc) main" | tee /etc/apt/sources.list.d/oxen.list && \
+    apt-get update && apt-get install -y \
+    python3-pyonionreq \
+    python3-oxenmq
```
I don't think this is used (I see it's also in the README, but probably was just copy and pasted from sogs).

mpretty-cyro replied:
This has been updated now

mpretty-cyro committed:
Addressed most PR comments

• Updated the database structure to allow for more extensive APIs in the future
• Moved the docker file, added a Readme and renamed it to make it clearer that it's for dev
• Cleaned up the auth verification logic
• Removed test data

mpretty-cyro committed:
Swapped from pyonionreq to session-util for onion req handling

jagerman committed:
make foreign keys deletion cascade

jagerman reviewed (approved):
Added one little db tweak to make foreign key deletions cascade, but aside from that this looks good to me now.
